joblib/parallel.py

```diff
@@ -692,65 +696,65 @@ def print_progress(self):
                 continue
-            # We need to be careful: the job list can be filling up as
-            # we empty it and Python list are not thread-safe by default hence
-            # the use of the lock
+
+            # We need to be careful: the buffer list can be filling up as
+            # we empty it and Python list are not thread-safe by default
+            # hence the use of the lock
             with self._lock:
                 job = self._jobs.pop(0)
+            if getattr(self._backend, 'supports_timeout', False):
+                res_batch = job.get(timeout=self.timeout)
+            else:
+                res_batch = job.get()
 
-            try:
-                if getattr(self._backend, 'supports_timeout', False):
-                    self._output.extend(job.get(timeout=self.timeout))
-                else:
-                    self._output.extend(job.get())
-
-            except BaseException as exception:
-                # Note: we catch any BaseException instead of just Exception
-                # instances to also include KeyboardInterrupt.
-
-                # Stop dispatching any new job in the async callback thread
-                self._aborting = True
-
-                # If the backend allows it, cancel or kill remaining running
-                # tasks without waiting for the results as we will raise
-                # the exception we got back to the caller instead of returning
-                # any result.
-                backend = self._backend
-                if (backend is not None and
-                        hasattr(backend, 'abort_everything')):
-                    # If the backend is managed externally we need to make sure
-                    # to leave it in a working state to allow for future jobs
-                    # scheduling.
-                    ensure_ready = self._managed_backend
-                    backend.abort_everything(ensure_ready=ensure_ready)
-
-                if not isinstance(exception, TransportableException):
-                    raise
-                else:
-                    # Capture exception to add information on the local
-                    # stack in addition to the distant stack
-                    this_report = format_outer_frames(context=10,
-                                                      stack_start=1)
-                    report = """Multiprocessing exception:
-%s
----------------------------------------------------------------------------
-Sub-process traceback:
----------------------------------------------------------------------------
-%s""" % (this_report, exception.message)
-                    # Convert this to a JoblibException
-                    exception_type = _mk_exception(exception.etype)[0]
-                    exception = exception_type(report)
-
-                    raise exception
+            for res in res_batch:
+                yield res
 
-    def __call__(self, iterable):
+    def _gen_results(self, iterable):
```

Comment:
In terms of naming / semantics, I think that we should revisit the "retrieve" / "_gen_results" function names. It is not clear to me what they do that is different.

Reply:
Let's keep only `retrieve` and merge the two.